### PR TITLE
[CARBONDATA-712]resolve the bug for bad records not written in csv file when 'BAD_RECORDS_ACTION'='REDIRECT'

### PR DESCRIPTION
**Pull request by anubhav100**
https://issues.apache.org/jira/browse/CARBONDATA-712

The bad_records are also not loading in the BAD_RECORS_LOCATION when using 'BAD_RECORDS_ACTION'='REDIRECT' at load time
both inprogress file and log file is empty

Because nothing in the I/O path guarantees that your data has reached disk,When you write data to a stream, it is not written immediately, and it is buffered. So we should use flush() when need to be sure that all your data from buffer is writte

so when we are using  bufferedCSVWriter.write(logStrings.toString()),it must be followed by
 bufferedCSVWriter.flush();

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/925/

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java [hunk 1]
```diff
@@ -179,6 +179,7 @@ private synchronized void writeBadRecordsToFile(StringBuilder logStrings) {
 
       }
       bufferedWriter.write(logStrings.toString());
+      bufferedWriter.flush();
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java]: It is not the correct way, please don't flush for every row. it slows down a lot.
 Please close the streams properly, it will solve the problem

**Comment by anubhav100** [comment]
@ravipesala changes done can you have a look

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/926/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/927/

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java (hunk 1 above)
Comment (anubhav100) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java]: Changes are done can u review?

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java [hunk 2]
```diff
@@ -176,15 +176,13 @@ private synchronized void writeBadRecordsToFile(StringBuilder logStrings) {
...
-
       }
       bufferedWriter.write(logStrings.toString());
-      bufferedWriter.flush();
       bufferedWriter.newLine();
     } catch (FileNotFoundException e) {
       LOGGER.error("Bad Log Files not found");
     } catch (IOException e) {
-      LOGGER.error("Error While writing bad log File");
+      LOGGER.error("Error While writ1ing bad log File");
```
Comment (chenliang613) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java]: "writ1ing" is wrong.

Comment (anubhav100) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java]: @chenliang613 i corrected it you can review?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/936/

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java [hunk 3]
```diff
@@ -176,7 +176,6 @@ private synchronized void writeBadRecordsToFile(StringBuilder logStrings) {
 
         bufferedWriter = new BufferedWriter(new OutputStreamWriter(outStream,
             Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET)));
-
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/BadRecordsLogger.java]: don't do unnecessary changes

**Comment by ravipesala** [comment]
@anubhav100  it is not correct way to close streams for every row, how the other rows will be added to bad records if it is closed for each row.

**Comment by anubhav100** [comment]
@ravipesala can you have another look?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/938/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/939/

**Comment by anubhav100** [comment]
@ravipesala can you review?

File: processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataConverterProcessorStepImpl.java [hunk 4]
```diff
@@ -93,6 +93,7 @@ protected CarbonRowBatch processRowBatch(CarbonRowBatch rowBatch, RowConverter l
     while (rowBatch.hasNext()) {
       newBatch.addRow(localConverter.convert(rowBatch.next()));
     }
+    createBadRecordLogger().closeStreams();
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataConverterProcessorStepImpl.java]: Here also we should not close badrecords after every batch process.

Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataConverterProcessorStepImpl.java]: Please call closeStreams from close method.

File: processing/src/main/java/org/apache/carbondata/processing/newflow/converter/impl/RowConverterImpl.java [hunk 5]
```diff
@@ -135,6 +135,7 @@ public CarbonRow convert(CarbonRow row) throws CarbonDataLoadingException {
           return null;
         }
       }
+
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/converter/impl/RowConverterImpl.java]: remove unncessary line

File: processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataConverterProcessorWithBucketingStepImpl.java [hunk 6]
```diff
@@ -120,6 +120,7 @@ protected CarbonRowBatch processRowBatch(CarbonRowBatch rowBatch, RowConverter l
       convertRow.bucketNumber = (short) partitioner.getPartition(next.getData());
       newBatch.addRow(convertRow);
     }
+    createBadRecordLogger().closeStreams();
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataConverterProcessorWithBucketingStepImpl.java]: Here also call from close method

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java [hunk 7]
```diff
@@ -950,6 +950,7 @@ private String getCarbonLocalBaseStoreLocation() {
     if (null == r[0] && badRecordsLogger.isBadRecordConvertNullDisable()) {
       badRecordsLogger
           .addBadRecordsToBuilder(r, "Column Names are coming NULL");
+      badRecordsLogger.closeStreams();
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: don't change anything in this class as it is already handled here.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/945/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/946/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/948/

**Comment by anubhav100** [comment]
@ravipesala can you review again?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/949/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/950/

**Comment by ravipesala** [comment]
LGTM
